index.bs

```diff
@@ -6754,6 +6754,14 @@ Note: this extension may be implemented for [=authenticators=] that do not use [
     dictionary AuthenticationExtensionsPRFInputs {
         AuthenticationExtensionsPRFValues eval;
         record<USVString, AuthenticationExtensionsPRFValues> evalByCredential;
+        AuthenticationExtensionsPRFImportCryptoKeyParams importCryptoKey;
+    };
+
+    dictionary AuthenticationExtensionsPRFImportCryptoKeyParams {
+        required KeyFormat format;
+        required AlgorithmIdentifier algorithm;
```

Comment:
Do we want to subset algorithms?

HMAC is useless, right? Because nobody else could have the HMAC key. (And there's HKDF if someone wants a PRF.)

RSA is out.

PBKDF2 doesn't make sense?

AES-* work.

ECDSA and ECDH? We could define things to work with P-256 if we wanted. Are asymmetric algorithms interesting? The ability to sign things as an identity without throwing UI? The ability to public-key decrypt? I guess this would imply bringing in [CryptoKeyPair](https://developer.mozilla.org/en-US/docs/Web/API/CryptoKeyPair).

Reply:
> Do we want to subset algorithms?

Maybe not? See: #1946

> HMAC is useless, right? Because nobody else could have the HMAC key. (And there's HKDF if someone wants a PRF.)

Mostly yes, unless you import it as extractable and use `wrapKey` to move it somewhere else (or `exportKey` of course, but that defeats the point of all this).

> PBKDF2 doesn't make sense?

Agreed.

> Are asymmetric algorithms interesting? The ability to sign things as an identity without throwing UI? The ability to public-key decrypt?

Signing is a bit awkward due to the lack of UI as you note, but asymmetric encryption certainly is interesting. For example, it would allow for encrypting to multiple PRF-backed keys for redundancy. This can be done today via key wrapping, but [not with never-extractable keys](https://github.com/w3c/webauthn/issues/1895#issuecomment-1691798283).

> ECDSA and ECDH? We could define things to work with P-256 if we wanted.

Right, we could. We can't naively use the PRF directly as a private key, because the P-256 key space is slightly smaller than the full 2<sup>32</sup>-1 range, so we'd need to specify some way to either fail out or recalculate some input salt in case the key falls outside the valid range. I would prefer if we could do something like that generically in WebCrypto, for example extend HKDF with asymmetric keys as output keys, but I presume it would take far more effort to get something like that accepted and rolled out in implementations.

The more I've thought about this, though, the more I've come to think that non-extractability in WebCrypto perhaps isn't that impressive after all. A malicious actor that has the ability to access `CryptoKey` objects probably also has the ability to overwrite properties on `window.crypto.subtle` - and therefore also has the ability to intercept any WebCrypto invocation, set `extractable` to `true` and exfiltrate the key. Even without that, just access to a `CryptoKey` object is enough to use that key for signing and decryption operations - you don't really need to steal the keys if you can just steal what they were meant to protect.

I still think there is value in the ability to import PRF outputs directly to `CryptoKey` objects, as that still reduces an honest application's exposure to raw key material. But I'm not sure that non-extractability is really that important - and if it's not, we might as well hard-code this to output HKDF keys and let application developers go from there. (WebCrypto HKDF cannot derive asymmetric keys, but you can generate asymmetric keys and wrap them with HKDF-derived keys.)

On the other hand, I guess this line of reasoning also comes back around in favour of importing PRF outputs directly to asymmetric private keys - since the domain separation here actually can make truly non-extractable keys. It's still true that malicious code could abuse the resulting `CryptoKey` object, but a truly non-extractable ECDH key is a powerful thing to have access to. Again in the multi-recipient encryption example, you could use ECDH to derive symmetric encryption keys - those `CryptoKey`s would be extractable, but they would exist only alongside the cleartext anyway. And with ECDH you could rotate the encryption keys after every decryption, so that any leaked key won't be able to access any future versions of the contents.

So okay, do you think it would make sense to specify a procedure for the client to transform the PRF output into an asymmetric key? I'm thinking in that case we should probably base it on HKDF or the like, rather than restricting ourselves to key sizes around 32 bits only.